buildSrc/plugins/src/main/kotlin/org/jetbrains/kotlin/KotlinNativeTest.kt

```diff
@@ -0,0 +1,436 @@
+
+abstract class KonanTest : DefaultTask() {
+    enum class Logger {
+        EMPTY,    // Built without test runner
+        GTEST,    // Google test log output
+        TEAMCITY, // TeamCity log output
+        SIMPLE,   // Prints simple messages of passed/failed tests
+        SILENT    // Prints no log of passed/failed tests
+    }
+
+    var disabled: Boolean
+        get() = !enabled
+        @Optional set(value) { enabled = !value }
+
+    /**
+     * Test output directory. Used to store processed sources and binary artifacts.
+     */
+    lateinit var outputDirectory: String
+
+    /**
+     * Test logger to be used for the test built with TestRunner (`-tr` option).
+     */
+    @Optional
+    lateinit var testLogger: Logger
+
+    /**
+     * Test executable arguments.
+     */
+    @Input
+    lateinit var arguments: MutableList<String>
+
+    /**
+     * Test executable.
+     */
+    @Input
+    lateinit var executable: String
+
+    /**
+     * Test source.
+     */
+    @Optional
+    lateinit var source: String
+
+    /**
+     * Sets test filtering to choose the exact test in the executable built with TestRunner.
+     */
+    @Input
+    var useFilter = true
+
+    /**
+     * An action to be executed before the build.
+     * As this run task comes after the build task all actions for doFirst
+     * should be done before the build and not run.
+     */
+    @Optional
+    var doBefore: Action<in Task>? = null
+
+    @Suppress("UnstableApiUsage")
+    override fun configure(config: Closure<*>): Task {
+        super.configure(config)
+
+        // Set Gradle properties for the better navigation
+        group = LifecycleBasePlugin.VERIFICATION_GROUP
+        description = "Kotlin/Native test infrastructure task"
+
+        if (!::arguments.isInitialized) {
+            arguments = mutableListOf()
+        }
+        if (::testLogger.isInitialized && testLogger != Logger.EMPTY) {
+            arguments.add("--ktest_logger=$testLogger")
+        }
+        if (useFilter && ::source.isInitialized) {
+            arguments.add("--ktest_filter=${source.convertToPattern()}")
+        }
+        project.setDistDependencyFor(this)
+        return this
+    }
+
+    @TaskAction
+    open fun run() = project.executeAndCheck(project.file(executable).toPath(), arguments)
+
+    // Converts to runner's pattern
+    private fun String.convertToPattern() = this.replace('/', '.').replace(".kt", "") + (".*")
+
+    internal fun ProcessOutput.print(prepend: String = "") {
+        if (project.verboseTest)
+            println(prepend + """
+                |stdout:$stdOut
+                |stderr:$stdErr
+                |exit code: $exitCode
+                """.trimMargin())
+    }
+}
+
+/**
+ * Create a test task of the given type. Supports configuration with Closure passed form build.gradle file.
+ */
+fun <T: KonanTest> Project.createTest(name: String, type: Class<T>, config: Closure<*>): T =
```

Comment:
It may make sense to add an overloaded method taking an `Action<T>` as a last parameter. In Gradle Kotlin DSL, a lambda can be passed as such an `Action` so this method will be useful in *.kts build scripts.